FindMKL.cmake

```diff
@@ -1,230 +1,308 @@
+
+  mkl::mkl_intel_32bit_omp_dyn
+
+BLACS targets:
+
+  mkl::blacs_[mpich|ompi]_[gf|intel]_[32bit|64bit]_[seq|omp|tbb]_[st|dyn] e.g.
+
+  mkl::blacs_intel_mpich_32bit_seq_st
+
+ScaLAPACK targets:
+
+  mkl::scalapack_[mpich|ompi]_[gf|intel]_[32bit|64bit]_[seq|omp|tbb]_[st|dyn] e.g.
 
-``mkl::scalapack``
+  mkl::scalapack_intel_mpich_64bit_omp_dyn
 
 Result variables
 ^^^^^^^^^^^^^^^^
 
-``MKL_FOUND``
-  Found MKL.
-
-``MKL_SCALAPACK_FOUND``
-  Found ScaLAPACK.
+MKL_FOUND
 
 Not supported
 ^^^^^^^^^^^^^
 
-- TBB threading back-end
 - F95 interfaces
 
-Note: Mixing GCC and Intel OpenMP backends is a bad idea.
-
 #]=======================================================================]
 
+cmake_minimum_required(VERSION 3.12)
+
 # Modules
 #
 include(FindPackageHandleStandardArgs)
 
-# Functions
-#
-function(__mkl_find_library _name)
-    find_library(${_name}
-        NAMES ${ARGN}
-        HINTS ${MKL_ROOT}
-        PATH_SUFFIXES ${_mkl_libpath_suffix}
-                      lib
-        )
-    mark_as_advanced(${_name})
-endfunction()
+if(NOT (CMAKE_C_COMPILER_LOADED OR
+        CMAKE_CXX_COMPILER_LOADED OR
+        CMAKE_Fortran_COMPILER_LOADED))
+    message(FATAL_ERROR "FindMKL requires Fortran, C, or C++ to be enabled.")
+endif()
 
-# Options
-#
-# The `NOT DEFINED` guards on CACHED variables are needed to make sure that 
-# normal variables of the same name always take precedence*.
-#
-# * There are many caveats with CACHE variables in CMake. Before version 
-#   3.12, both `option()` and `set(... CACHE ...)` would override normal 
-#   variables if cached equivalents don't exist or they exisit but their type 
-#   is not specified (e.g. command line arguments: -DFOO=ON instead of 
-#   -DFOO:BOOL=ON). For 3.13 with policy CMP0077, `option()` no longer overrides 
-#   normal variables of the same name. `set(... CACHE ...)` is still stuck with 
-#   the old behaviour. 
+# Dependencies
 #
-#   https://cmake.org/cmake/help/v3.15/command/set.html#set-cache-entry
-#   https://cmake.org/cmake/help/v3.15/policy/CMP0077.html
+find_package(Threads)
+find_package(MPI COMPONENTS CXX)
+find_package(OpenMP COMPONENTS CXX)
+
+# If MKL_ROOT is not set, set it via the env variable MKLROOT.
 #
 if(NOT DEFINED MKL_ROOT)
     set(MKL_ROOT $ENV{MKLROOT} CACHE PATH "MKL's root directory.")
 endif()
 
-if(NOT DEFINED MKL_PARALLEL)
-    set(MKL_PARALLEL ON CACHE BOOL "Toggles parallel and sequential versions.")
-endif()
-
-if(NOT DEFINED MKL_64BIT)
-    set(MKL_64BIT OFF CACHE BOOL "Toggles 32bit/64bit MKL interfaces.")
-endif()
-
 # Determine MKL's library folder
 #
 set(_mkl_libpath_suffix "lib/intel64")
 if(CMAKE_SIZEOF_VOID_P EQUAL 4) # 32 bit
     set(_mkl_libpath_suffix "lib/ia32")
 endif()
 
+set(_mkl_libname_prefix "lib")
+set(_mkl_shared_lib ".so")
+set(_mkl_static_lib ".a")
 if(WIN32)
     string(APPEND _mkl_libpath_suffix "_win")
+    set(_mkl_libname_prefix "")
+    set(_mkl_shared_lib "_dll.lib")
+    set(_mkl_static_lib ".lib")
 elseif(APPLE)
```

Comment:
On macOS, shared libraries have the ".dylib" suffix, so one should probably add set(_mkl_shared_lib ".dylib") here.